### PR TITLE
[CB] Add validation mode to Sampler

### PR DESCRIPTION
**Pull request by iefode**
Tickets:
* [153600](https://jira.devtools.intel.com/browse/CVS-153600)
* [153601](https://jira.devtools.intel.com/browse/CVS-153601)

### CODE REVIEW

File: src/cpp/src/sampler.cpp [hunk 1]
```diff
@@ -587,45 +590,95 @@ SamplerOutput Sampler::sample(std::vector<SequenceGroup::Ptr> & sequence_groups,
 
         const void * sequence_group_logits_data = logits_data + vocab_size * currently_processed_tokens;
         ov::Tensor sequence_group_logits(ov::element::f32, ov::Shape{num_running_sequences, actual_seq_len, vocab_size}, (void *)sequence_group_logits_data);
-
+        size_t decrease_len = 0;
         if (sequence_group->requires_sampling()) {
+            auto token_id = actual_seq_len;
+            // prompt phase
+            if (sequence_group->get_num_processed_tokens() == 0) {
```
Comment (ilya-lavrenov) [inline src/cpp/src/sampler.cpp:597]: in case of chunked prefill, prompt can be processed via multiple iterations.
And it seems we will have overflow below after subtraction

File: src/cpp/src/sampler.cpp [hunk 2]
```diff
@@ -587,45 +590,95 @@ SamplerOutput Sampler::sample(std::vector<SequenceGroup::Ptr> & sequence_groups,
...
+                token_id -= (sequence_group->get_prompt_len() - 1);
+            }
             if (sampling_params.is_greedy_decoding() || sampling_params.is_multinomial()) {
                 std::vector<Sequence::Ptr> running_sequences = sequence_group->get_running_sequences();
-                if (sampling_params.is_greedy_decoding()) {
-                    OPENVINO_ASSERT(num_running_sequences == 1);
+                // limitation for speculative decoding
+                if (is_validation_mode_enabled || sampling_params.is_greedy_decoding()) {
+                    OPENVINO_ASSERT(running_sequences.size() == 1);
```
Comment (ilya-lavrenov) [inline src/cpp/src/sampler.cpp:604]: speculative decoding should work with parallel sampling as well, right? in a theory

**Comment by iefode** [comment]
@ilya-lavrenov Could you please take a look this PR?

File: src/cpp/src/sampler.cpp [hunk 3]
```diff
@@ -587,45 +590,95 @@ SamplerOutput Sampler::sample(std::vector<SequenceGroup::Ptr> & sequence_groups,
...
+                                    sampler_output.m_forked_sequences.insert({running_sequences[0]->get_id(), forked_seq_ids});
+                                }
+                            }
+                            // flag to add sampled token to generated sequence or extend logit processors only
+                            bool is_extend_sequence = token_id_per_seq == 0;
+                            if (token_id_per_seq > 0) {
+                                auto it = running_seq_token_ids.rbegin();
+                                std::advance(it, token_id_per_seq - 1);
+                                // to validate candidates from assisting model and remove incorrect ones from generated sequence
+                                if (is_validation_mode_enabled && *it != sampled_token_id.m_index) {
```
Comment (ilya-lavrenov) [inline src/cpp/src/sampler.cpp:666]: in general, the sampler logic become even more complex as now it's responsible for validation (part of speculative decoding).

Is it possible to implement speculative part on top of existing sampler functionality (which is just to sampler new tokens w/o any verification, removing tokens from end of sequences) ?

Maybe we need to split implementation of current sampler into parts:
- provide next token based on logits tensor - basic building block, which does not depend on sequence group. I suppose this building block can be also reused in LLM stateful pipeline, Visual language and Whisper pipelines.
- then, adding this new sampled token to sequence group. This bullet implies that output of sampler is consumed by sequence group as is.

In this case, speculative sampler logic can be implemented on top of building blocks from bullet 1 and "basic" sampler (bullet 2) will not be "overcompilcated" by speculative logic

**Review by ilya-lavrenov — CHANGES_REQUESTED** [review]

File: src/cpp/src/sampler.cpp (hunk 3 above)
Comment (iefode) [reply src/cpp/src/sampler.cpp]: I have separated some functionality from `sample` method to stand alone functions defined in the cpp file. The changes allows to simplify the method and improve readability. Please check

File: src/cpp/src/sampler.cpp (hunk 2 above)
Comment (iefode) [reply src/cpp/src/sampler.cpp:671]: This limitation was introduced because we were not sure how to handle speculative decoding with several sequences in case of different match number. I have remove this assert. But my suggestion is to check this case in real scenario and return this condition in case of unexpected behavior with TODO and link to the ticket

File: src/cpp/src/sampler.cpp (hunk 1 above)
Comment (iefode) [reply src/cpp/src/sampler.cpp]: It's handled inside the condition. I have updated the code just to improve readability

**Comment by iefode** [comment]
@ilya-lavrenov Please take a look

File: src/cpp/src/sampler.cpp [hunk 4]
```diff
@@ -587,45 +662,67 @@ SamplerOutput Sampler::sample(std::vector<SequenceGroup::Ptr> & sequence_groups,
...
+                        bool is_extend_sequence = token_offset == 0,
+                             // flag to update generated length of sequence group in logit processor
+                             is_update_len_logit_processor = running_sequence_id == num_running_sequences - 1,
+                             is_validation_failed = true;
+                        if (is_validation_mode_enabled) {
+                            is_validation_failed = validate_candidate(running_sequences[running_sequence_id], token_offset, sampled_token_id, is_extend_sequence, decrease_context_len_per_seq_group);
+                        }
+                        register_new_token(sampled_token_id, running_sequences[running_sequence_id], logit_processor, is_extend_sequence, is_update_len_logit_processor);
+                        // to exit from sampling in case of failed token validation
+                        if (!is_validation_failed) {
```
Comment (ilya-lavrenov) [inline src/cpp/src/sampler.cpp:721]:
```suggestion
                        if (is_validation_failed) {
```
should it be corrected like this?

File: src/cpp/src/sequence_group.hpp [hunk 5]
```diff
@@ -413,18 +420,29 @@ class SequenceGroup {
...
         return m_num_scheduled_tokens > 0;
     }
 
+    void set_num_validated_tokens(size_t k) {
+        // in case of non-prompt we need to take prev tokens + token to validate
+        // m_num_validated_tokens = get_num_processed_tokens() ? k + 1 : k;
+        m_num_validated_tokens = k;
+    }
+
+    size_t get_num_validated_tokens() {
```
Comment (ilya-lavrenov) [inline src/cpp/src/sequence_group.hpp:436]: validated => validation ? or get_num_tokens_to_validate ?
because validated means that they are confirmed by main models.. 

If you are OK, let's change validat**ed** in other places too

File: src/cpp/src/sampler.cpp [hunk 6]
```diff
@@ -560,15 +561,89 @@ std::vector<int64_t> Sampler::_try_finish_generation(SequenceGroup::Ptr & sequen
...
+                   Token& sampled_token,
+                   bool& is_extend_sequence,
+                   size_t& decrease_context_len_per_seq_group) {
+    if (token_idx > 0) {
+        const auto& generated_tokens = running_sequence->get_generated_ids();
+        auto it = generated_tokens.rbegin();
+        std::advance(it, token_idx - 1);
+        // to validate candidates from assisting model and remove incorrect ones from generated sequence
+        if (*it != sampled_token.m_index) {
+            running_sequence->remove_last_tokens(token_idx);
```
Comment (ilya-lavrenov) [inline src/cpp/src/sampler.cpp:626]: do you think we can also update `context_len` inside this method instead of calling `update_processed_tokens_num` separately ?

File: src/cpp/src/sampler.cpp [hunk 7]
```diff
@@ -658,14 +755,29 @@ SamplerOutput Sampler::sample(std::vector<SequenceGroup::Ptr> & sequence_groups,
...
+        if (decrease_context_len_per_seq_group) {
+            const auto num_processed_tokens = sequence_group->get_num_processed_tokens();
+            OPENVINO_ASSERT(num_processed_tokens >= decrease_context_len_per_seq_group);
+            OPENVINO_ASSERT(sequence_group->get_context_len() >= decrease_context_len_per_seq_group);
+            sequence_group->update_processed_tokens_num(num_processed_tokens - decrease_context_len_per_seq_group);
+        }
 
         // accumulate a number of processed tokens
-        currently_processed_tokens += padded_amount_of_processed_tokens * num_running_sequences;
+        currently_processed_tokens += (padded_amount_of_processed_tokens - decrease_context_len_per_seq_group) * num_running_sequences;
```
Comment (ilya-lavrenov) [inline src/cpp/src/sampler.cpp:767]: why do we need to subtract `decrease_context_len_per_seq_group` here? 
We are iterating over already computed logits tensor and looks like validation passed / failed does not make sense here

File: src/cpp/src/sampler.cpp (hunk 6 above)
Comment (iefode) [reply src/cpp/src/sampler.cpp:626]: It is done  per request by 
```
        // decrease sequence_group context in case of candidates generated by draft_model were not accepted by main_model
        if (decrease_context_len_per_seq_group) {
            const auto num_processed_tokens = sequence_group->get_num_processed_tokens();
            OPENVINO_ASSERT(num_processed_tokens >= decrease_context_len_per_seq_group);
            OPENVINO_ASSERT(sequence_group->get_context_len() >= decrease_context_len_per_seq_group);
            sequence_group->update_processed_tokens_num(num_processed_tokens - decrease_context_len_per_seq_group);
        }
```

File: src/cpp/src/sampler.cpp (hunk 7 above)
Comment (iefode) [reply src/cpp/src/sampler.cpp:767]: to get a correct offset here
```
const void * sequence_group_logits_data = logits_data + vocab_size * currently_processed_tokens;
```

File: src/cpp/src/sampler.cpp (hunk 4 above)
Comment (iefode) [reply src/cpp/src/sampler.cpp]: fixed

File: src/cpp/src/sequence_group.hpp (hunk 5 above)
Comment (iefode) [reply src/cpp/src/sequence_group.hpp]: applied

**Review by ilya-lavrenov — APPROVED** [review]

File: src/cpp/src/sampler.cpp (hunk 7 above)
Comment (ilya-lavrenov) [reply src/cpp/src/sampler.cpp:767]: agreed to fix separately
